### PR TITLE
Trapdoor crash fix

### PR DESCRIPTION
**Pull request by theophriene**
* Fixed trapdoor crashing server due to failed assert
* Added trapdoor rotation upon placement

I'm not familiar enough with api - most likely there's a better way to do things.
Awaiting your suggestions.

~~P.S.: Trpdoors still do move up in one tick after placement / on block update.~~ Not anymore

Also, comments in the code will be removed anyway

### CODE REVIEW

**Comment by theophriene** [comment]
Okay, seems to be non-compliant to vanilla. Back to testing

**Comment by theophriene** [comment]
Seems done.
Tested placing on horizontal surface, from the position above/ under the lapis block:

**Comment by theophriene** [comment]
> Seems done.
> Tested placing on horizontal surface, from the position above/ under the lapis block: (all the doors are opened afterwards to see the alignment)
> 
> 

The only issue is that the minecraft is extremely sensitive to position, causing things like this, when player is not aligned properly

**Review by bearbin — CHANGES_REQUESTED** [review]
I don't like calculating a 'safe block face', and just calculating block meta using that face. 

I would replace `BlockFaceToMetaData` with a new function `BlockFaceAndOffsetToMetaData`, replacing the existing BlockFaceToMetaData function.

Also, please don't use C-style casts and instead use C++-style.

Edit: in terms of behaviour, this PR looks good to me. It may not be exactly vanilla behaviour but it's sensical enough.

**Comment by bearbin** [comment]
Also, please add yourself to the CONTRIBUTORS file to accept the licensing terms and be recognised as a Cuberite contributor.

**Comment by theophriene** [comment]
> Also, please add yourself to the CONTRIBUTORS file to accept the licensing terms and be recognised as a Cuberite contributor.

Will do, when code is ready quality-wise

**Review by bearbin — APPROVED** [review]

File: src/Blocks/BlockTrapdoor.h [hunk 1]
```diff
@@ -84,31 +84,72 @@ class cBlockTrapdoorHandler :
...
 		a_BlockType = m_BlockType;
-		a_BlockMeta = BlockFaceToMetaData(a_ClickedBlockFace);
+		a_BlockMeta = BlockFaceAndOffsetToMetaData(a_ClickedBlockFace, BlockOffsetRelToPlayer);
 
-		if (a_CursorPos.y > 7)
+		// Trapdoor is placed on top of a block
+		if ((a_ClickedBlockFace == BLOCK_FACE_YP))
+		{
+			// Toggle 'Move up half-block' bit off
+			a_BlockMeta &= a_BlockMeta | 0x8;
```
Comment (peterbell10) [inline src/Blocks/BlockTrapdoor.h:98]: This line doesn't do anything (x & (x | y) is just x). Perhaps you meant
```cpp
a_BlockMeta &= ~0x08;
```
which unsets the 8s bit.

**Review by peterbell10 — DISMISSED** [review]

File: src/Blocks/BlockTrapdoor.h (hunk 1 above)
Comment (theophriene) [reply src/Blocks/BlockTrapdoor.h]: fixed.

**Comment by tigerw** [comment]
On YM/YP scenarios this places the trapdoor with the hinge dependant on the player's position relative to the block, instead vanilla puts the hinge dependant on the player's look vector. Let me try to apply a change

**Review by tigerw — APPROVED** [review]

**Comment by tigerw** [comment]
@theophriene pushed some changes for your consideration

**Comment by theophriene** [comment]
Oh, wow! That's what I thought should exist somewhere. I should've checked the base class from the start. Thanks.

**Review by bearbin — APPROVED** [review]

**Pull request merged**